Refactoring ini parsing code into strategy pattern and adding unit tests

mcgoober opened the pull request:
Whilst attempting to create an ini file editor in LancerEdit I hit on a number of issues with the current interface to ini file parsing. This change prepares the foundations for such an editor, but also allows for unit tests of the ini parsers. 
A number of issues with the binary parser have been addressed and a set of potential issues with the text parser have been identified.

CallumDev commented:
Hi, a couple of issues with this: 

- Exceptions are not allowed for INI parsing errors. While I'm happy for you to refactor the error reporting so it can be used outside of being dumped to disk, the correct behaviour for the engine is to log errors and try best effort to read values skipping over junk. Mod compatibility breaks every time INI parsing behaviour is tightened

- I've noticed some references to a "binitools" repository. I don't think it has been verified against the original INI_reader from FL's common.dll behaviour which is very permissive, and I would not be using it as a reference. Due to mods being the way they are, we need to be permissive with input as much as possible.

Particularly I'm not sure if the string quoting syntax actually exists in Freelancer's dialect of ini. Any changes in the ini parsing code really need to be verified against common.dll's INI_Reader class.

mcgoober commented:
Many thanks for your speedy comments, yep keep forgetting about FL mods.
Do we happen to have a decompiled version of common.dll INI_Reader class? was the C# parsing code based on it?
Do you have any pointers to particularly troublesome ini files?

CallumDev commented:
I've set up a simple test case using Freelancer's dlls here:

https://github.com/Librelancer/flini-reader-test

Currently only compiles on linux but should be adaptable to MSVC easily (may compile out of the box ymmv).

The C# code is based simply on community observations, and then wildly optimised, not based on `INI_Reader` in the slightest. I would not bother decompiling it as the paradigm is entirely different, and I don't want to run into legal troubles there.

From quickly testing the FL ini parser two things are apparent:

- `[Header` will start a section called `Header`, I am unsure if this works correctly in Librelancer - but the test cases should reflect that this is an accepted input.
- As suspected, quoting does not trigger any special behaviour - the string is copied verbatim.

Hope this helps

mcgoober commented:
Hey @CallumDev, I've been playing with the fini_reader_test code. This has been really interesting and I'm keen to reveal what I've been working on. Whilst conducting the investigation I found that I also need to make changes to the value converters as I've observed some differences there also between the original game and this code. 
Therefore, to get these changes over the line I've removed the extra exception throwing code from the pull request and added in a bit of refactoring for the value converters to allow me to make a more compatible version of the ini parser without breaking the existing implementation.

CallumDev commented on src/LibreLancer.Tests/Ini/TextIniParserTests.cs:
```diff
@@ -0,0 +1,203 @@
+            entry[3].ToString().Should().Be("\"+1\"");
+        }
+
+        /* TODO: Failing Tests:
```
Remove block

CallumDev commented on src/LibreLancer.Tests/Ini/TextIniParserTests.cs:
```diff
@@ -0,0 +1,203 @@
+        }
+
+
+        // https://github.com/skeeto/binitools?tab=readme-ov-file#text-format
```
Remove binitools reference, can lead astray

CallumDev reviewed (changes requested):
Happy to merge once the binitools comments are cleaned up
